### PR TITLE
[CPU EP] Add block quantized Gather contrib op

### PR DESCRIPTION
**Pull request by fajin-corp**
### Description
Add a gather that supports block-quantized input data.

### Motivation and Context
To support Web inference scenario with quantized vocabulary embeddings.

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 1]
```diff
@@ -3544,6 +3544,124 @@ MatMulBnb4 is a MatMul with weight quantized with 4 bits using either FP4 or NF4
...
+      .Input(0, "data", "Tensor of rank r >= 1. Block-wise quantized.", "T1")
+      .Input(1,
+             "indices",
+             "Tensor of int32/int64 indices, of any rank q. All index values are expected to be within bounds [-s, s-1] "
+             "along axis of size s. It is an error if any of the index values are out of bounds.",
+             "Tind")
+      .Input(2, "scales", "quantization scale", "T2")
+      .Input(3, "zero_points", "quantization zero points", "T1", OpSchema::Optional)
+      .Output(0, "output", "Dequantized output tensor of rank q + (r - 1).", "T2")
+      .TypeConstraint("T1", {"tensor(int4)", "tensor(uint4)"}, "Constrain quantized types.")
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:3583]: We can extend to 8bit in a separate following PR.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -3544,6 +3544,124 @@
...
+      .TypeConstraint("T2", {"tensor(float)", "tensor(float16)", "tensor(bfloat16)"}, "Constrain dequantized types.")
+      .TypeConstraint("Tind", {"tensor(int32)", "tensor(int64)"}, "Constrain indices to integer types.")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        // Type inference
+        propagateElemTypeFromInputToOutput(ctx, 2, 0);
+        // Shape inference
+        if (!hasNInputShapes(ctx, 3)) {
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:3590]: >3 [](http://example.com/codeflow?start=34&length=1)

it should be 2. The parameter is the max index id to check, not the count of input.

File: onnxruntime/contrib_ops/cpu/quantization/gather_block_quantized.cc [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+  // TODO(fajin): use SIMD
+  for (int64_t i = 0; i < N; ++i) {
+    auto idx = static_cast<int64_t>(indices_data[i]);
+    if (idx < -gather_axis_dim || idx >= gather_axis_dim) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "indices element out of data bounds, idx=", idx,
+                             " must be within the inclusive range [", -gather_axis_dim, ",", gather_axis_dim - 1, "]");
+    }
+  }
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/quantization/gather_block_quantized.cc:137]: Can we validate this together with the data retrieve.

Comment (fajin-corp) [reply onnxruntime/contrib_ops/cpu/quantization/gather_block_quantized.cc:137]: moved to data retrive code in next iter

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 2 above)
Comment (fajin-corp) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:3590]: onnx seems to have changed the implementation. now n stands for the number of input to check.

**Review by yufenglee — APPROVED** [review]

**Pull request merged**